File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/DataObjectVisitors.java
/*
 * Copyright (c) 2019 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 */
package org.eclipse.scout.rt.dataobject;

import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;

public final class DataObjectVisitors {

  private DataObjectVisitors() {
  }

  /**
   * Visits all nodes and calls on each element of given type the provided consumer. Matching nodes are visited
   * <b>not</b> recursively. If a node matches child nodes of this node are not visited.
   */
  public static <T> void forEach(Object root, Class<? extends T> elementType, Consumer<T> consumer) {
    forEach(root, new P_TypedElementConsumer<>(elementType, consumer, false));
  }

  /**
   * Visits all nodes and calls on each element of given type the provided consumer. Matching nodes are visited
   * recursively. If a node matches, child nodes of this node are also visited.
   */
  public static <T> void forEachRec(Object root, Class<? extends T> elementType, Consumer<T> consumer) {
    forEach(root, new P_TypedElementConsumer<>(elementType, consumer, true));
  }

  /**
   * Visits all nodes and calls on each element of given type the provided consumer. If consumer returns {@code true}
   * then matching node is visited recursively.
   */
  public static <T> void forEachRecIf(Object root, Class<? extends T> elementType, Predicate<T> consumer) {
    forEach(root, new P_TypedElementConsumer<>(elementType, consumer));
  }

  private static void forEach(Object root, Predicate<Object> elementConsumer) {
    new P_DataObjectVisitor(elementConsumer).visit(root);
  }

  /**
   * Visits all nodes and calls on each element of given type the provided operator. The current value will be replace
   * with the new value from calling the operator. Children of nodes that match element type are not visited.
   * <p>
   * Note: this operation might fail if data object contains unmodifiable collections or maps.
   */
  public static <T> void replaceEach(Object root, Class<? extends T> elementType, UnaryOperator<T> operator) {
    new P_ReplaceDataObjectVisitor<>(elementType, operator).visit(root);
  }

  private static final class P_DataObjectVisitor extends AbstractDataObjectVisitor {

    private final Predicate<Object> m_elementConsumer;

    private P_DataObjectVisitor(Predicate<Object> elementConsumer) {
      m_elementConsumer = elementConsumer;
    }

    @Override
    protected <T> void caseNode(T node, Consumer<T> chain) {
      if (m_elementConsumer.test(node)) {
        super.caseNode(node, chain); // call chain
      }
    }
  }

  private static final class P_TypedElementConsumer<T> implements Predicate<Object> {

    private final Class<? extends T> m_elementType;
    private final Predicate<T> m_elementConsumer;

    private P_TypedElementConsumer(Class<? extends T> elementType, Consumer<T> elementConsumer, boolean recursive) {
      this(elementType, o -> {
        elementConsumer.accept(o);
        return recursive;
      });
    }

    private P_TypedElementConsumer(Class<? extends T> elementType, Predicate<T> elementConsumer) {
      m_elementType = elementType;
      m_elementConsumer = elementConsumer;
    }

    @Override
    public boolean test(Object t) {
      if (m_elementType.isInstance(t)) {
        return m_elementConsumer.test(m_elementType.cast(t));
      }
      return true;
    }
  }

  private static final class P_ReplaceDataObjectVisitor<T> extends AbstractReplacingDataObjectVisitor {

    private final Class<? extends T> m_elementType;
    private final UnaryOperator<T> m_operator;

    public P_ReplaceDataObjectVisitor(Class<? extends T> elementType, UnaryOperator<T> operator) {
      m_elementType = elementType;
      m_operator = operator;
    }

    @Override
    protected <OT> OT replaceOrVisit(OT o) {
      if (m_elementType.isInstance(o)) {
        //noinspection unchecked
        return (OT) m_operator.apply(m_elementType.cast(o));
      }
      else {
        return super.replaceOrVisit(o);
      }
    }
  }
}

File: eclipse-scout-core/src/dataobject/dataObjectVisitors.ts
/*
 * Copyright (c) 2010, 2026 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */

import {arrays, BaseDoEntity, Constructor, objects, scout, strings, TreeVisitResult} from '../index';

export const dataObjectVisitors = {

  /**
   * Visits all nodes and calls on each element matching the given type the provided consumer.
   * If a node matches, child nodes of this node are not visited.
   */
  forEach<T>(root: any, type: Constructor<T>, consumer: (node: T) => void) {
    dataObjectVisitors.forEachIf(root, (type ? c => c instanceof type : c => true) as (candidate) => candidate is T, consumer);
  },

  /**
   * Visits all nodes and calls on each element matching the given type the provided consumer.
   * If a node matches, child nodes of this node are also visited.
   */
  forEachRec<T>(root: any, type: Constructor<T>, consumer: (node: T) => void) {
    dataObjectVisitors.forEachIfRec(root, (type ? c => c instanceof type : c => true) as (candidate) => candidate is T, consumer);
  },

  /**
   * Visits all nodes and calls on each element matching the given type the provided visitor.
   * If a node matches, the resulting {@link TreeVisitResult} of the given visitor determines how the visitor proceeds.
   */
  forEachRecWhile<T>(root: any, type: Constructor<T>, visitor: DataObjectVisitor<T>) {
    dataObjectVisitors.forEachIfRecWhile(root, (type ? c => c instanceof type : c => true) as (candidate) => candidate is T, visitor);
  },

  /**
   * Visits all nodes and calls on each element matching the given type predicate the provided consumer.
   * If a node matches, child nodes of this node are not visited.
   */
  forEachIf<T>(root: any, typePredicate: (candidate) => candidate is T, consumer: (node: T) => void) {
    dataObjectVisitors.forEachIfRecWhile(root, typePredicate, node => {
      consumer(node);
      return TreeVisitResult.SKIP_SUBTREE;
    });
  },

  /**
   * Visits all nodes and calls on each element matching the given type predicate the provided consumer.
   * If a node matches, child nodes of this node are also visited.
   */
  forEachIfRec<T>(root: any, typePredicate: (candidate) => candidate is T, consumer: (node: T) => void) {
    dataObjectVisitors.forEachIfRecWhile(root, typePredicate, node => {
      consumer(node);
      return TreeVisitResult.CONTINUE;
    });
  },

  /**
   * Visits all nodes and calls on each element matching the given type predicate the provided visitor.
   * If a node matches, the resulting {@link TreeVisitResult} of the given visitor determines how the visitor proceeds.
   */
  forEachIfRecWhile<T>(root: any, typePredicate: (candidate) => candidate is T, visitor: DataObjectVisitor<T>) {
    if (!typePredicate) {
      typePredicate = (c => true) as (candidate) => candidate is T;
    }
    scout.create(DataObjectVisitHelper).visit(root, (node, context) => {
      if (typePredicate(node)) {
        return visitor(node, context);
      }
      return TreeVisitResult.CONTINUE;
    });
  }
};

/**
 * Helper class to visit data objects (depth first, pre-order).
 * <p>
 * {@link dataObjectVisitors} will be in most cases sufficient.
 */
export class DataObjectVisitHelper {

  /**
   * Visits the given node and all of its children using the visitor.
   */
  visit(node: any, visitor: DataObjectVisitor<any>) {
    this._visit(node, scout.create(DataObjectVisitorContext), visitor);
  }

  protected _visit(node: any, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    if (objects.isNullOrUndefined(node)) {
      return TreeVisitResult.CONTINUE;
    }

    if (node instanceof Set || objects.isArray(node)) {
      return this._visitNode(node, context, visitor, (n, c, v) => this._visitSetOrArray(n, c, v));
    }

    if (node instanceof Map) {
      return this._visitNode(node, context, visitor, (n, c, v) => this._visitMap(n, c, v));
    }

    if (node instanceof BaseDoEntity) {
      return this._visitNode(node, context, visitor, (n, c, v) => this._visitBaseDoEntity(n, c, v));
    }

    if (objects.isObject(node)) {
      return this._visitNode(node, context, visitor, (n, c, v) => this._visitObject(n, c, v));
    }

    return this._visitNode(node, context, visitor, (n, c, v) => this._visitAny(n, c, v));
  }

  /**
   * Visits the given node using the visitor.
   * If this results in {@link TreeVisitResult.TERMINATE} visiting is terminated.
   * If it does not result in {@link TreeVisitResult.SKIP_SUBTREE} the children of the given node are visited using the given chain.
   */
  protected _visitNode<T>(node: T, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>, chain: (n: T, c: DataObjectVisitorContext, v: DataObjectVisitor<any>) => TreeVisitResult): TreeVisitResult {
    if (context.containsParent(node)) {
      throw new Error('Unable to visit node. Reference cycle detected.');
    }

    const visitResult = visitor(node, context);
    if (visitResult === TreeVisitResult.TERMINATE) {
      return TreeVisitResult.TERMINATE;
    }

    if (visitResult !== TreeVisitResult.SKIP_SUBTREE) {
      context.pushParent(node);
      const chainVisitResult = chain(node, context, visitor);
      context.popParent();
      return chainVisitResult;
    }

    return TreeVisitResult.CONTINUE;
  }

  /**
   * Visits all elements of a {@link Set} or an {@link Array} using the visitor.
   */
  protected _visitSetOrArray(setOrArray: Set<any> | Array<any>, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    for (const element of setOrArray) {
      if (this._visit(element, context, visitor) === TreeVisitResult.TERMINATE) {
        return TreeVisitResult.TERMINATE;
      }
    }
    return TreeVisitResult.CONTINUE;
  }

  /**
   * Visits all keys and values of a {@link Map} using the visitor.
   */
  protected _visitMap(map: Map<any, any>, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    for (const [key, value] of map) {
      if (this._visit(key, context, visitor) === TreeVisitResult.TERMINATE) {
        return TreeVisitResult.TERMINATE;
      }
      if (this._visit(value, context, visitor) === TreeVisitResult.TERMINATE) {
        return TreeVisitResult.TERMINATE;
      }
    }
    return TreeVisitResult.CONTINUE;
  }

  /**
   * Visits all values and all contributions of a {@link BaseDoEntity} using the visitor.
   */
  protected _visitBaseDoEntity(baseDoEntity: BaseDoEntity, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    if (this._visitSetOrArray(
      Object
        .keys(baseDoEntity)
        .filter(key => !strings.startsWith(key, '_'))
        .map(key => baseDoEntity[key]),
      context,
      visitor
    ) === TreeVisitResult.TERMINATE) {
      return TreeVisitResult.TERMINATE;
    }
    return this._visitSetOrArray(baseDoEntity.getContributions(), context, visitor);
  }

  /**
   * Visits all values of an {@link object} using the visitor.
   */
  protected _visitObject(obj: object, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    return this._visitSetOrArray(Object.values(obj), context, visitor);
  }

  /**
   * Visits no child of the given element. Hook for subclasses.
   */
  protected _visitAny(node: any, context: DataObjectVisitorContext, visitor: DataObjectVisitor<any>): TreeVisitResult {
    return TreeVisitResult.CONTINUE;
  }
}

export type DataObjectVisitor<T> = (node: T, context: DataObjectVisitorContext) => TreeVisitResult;

export class DataObjectVisitorContext {

  protected _parents = [];

  get parents(): any[] {
    return [...this._parents];
  }

  parent(index = 0): any {
    return this._parents[this._parents.length - 1 - index];
  }

  containsParent(parent: any): boolean {
    return arrays.contains(this._parents, parent);
  }

  pushParent(parent: any) {
    this._parents.push(parent);
  }

  popParent(): any {
    return this._parents.pop();
  }
}
